File: apps/client-macos/Sources/TimeTrack/Activity/ActivityRateMeter.swift
import Foundation

/// Pure activity-% computation (PRD §6.3, Slice 2.3b): the 60s interval is split into N sub-buckets;
/// a bucket is "active" if any input (key OR pointer) occurred in it. activityPct = active/N × 100.
/// Separated from timing so it is fully unit-testable. Buckets beyond `buckets` are ignored.
struct ActivityRateMeter {
    private let buckets: Int
    private var seen = 0
    private var active = 0

    init(buckets: Int) { self.buckets = max(1, buckets) }

    mutating func addBucket(delta: (keys: UInt64, pointer: UInt64)) {
        guard seen < buckets else { return }
        seen += 1
        if delta.keys > 0 || delta.pointer > 0 { active += 1 }
    }

    /// round(active / buckets × 100), clamped 0…100.
    func activityPct() -> Int {
        let pct = (Double(active) / Double(buckets) * 100).rounded()
        return min(100, max(0, Int(pct)))
    }
}

File: apps/client-windows/src/NiftyTimer/Tracking/ActivityRateMeter.cs
namespace NiftyTimer.Tracking;

/// <summary>
/// PRD §6.3 — the pure activity-% computation. The sampling interval is split into N sub-buckets;
/// a bucket counts as "active" if ANY input occurred in it, key or pointer. The result is
/// <c>active / N × 100</c>.
///
/// Deliberately separated from timing so it is fully unit-testable, and deliberately NOT in a
/// capture namespace: it takes two integers per bucket and returns a percentage. It touches no
/// hardware and needs no gate — it sits beside <see cref="IdleMonitor"/>, the other pure machine
/// the capture layer drives.
///
/// A class rather than a struct: the Swift original is a mutable struct, and the C# equivalent
/// would be copied by every assignment and parameter pass, silently discarding buckets.
/// </summary>
public sealed class ActivityRateMeter
{
    private readonly int _buckets;
    private int _seen;
    private int _active;

    public ActivityRateMeter(int buckets) => _buckets = Math.Max(1, buckets);

    /// <summary>
    /// Record one sub-bucket from the input-count deltas measured across it. Buckets beyond the
    /// configured count are ignored, so a cycle that somehow over-runs cannot report above 100%.
    /// </summary>
    public void AddBucket(long keyDelta, long pointerDelta)
    {
        if (_seen >= _buckets)
        {
            return;
        }

        _seen++;
        if (keyDelta > 0 || pointerDelta > 0)
        {
            _active++;
        }
    }

    /// <summary>round(active / buckets × 100), clamped to 0…100 (the server's schema bound).</summary>
    public int ActivityPct()
    {
        var pct = (int)Math.Round(_active / (double)_buckets * 100, MidpointRounding.AwayFromZero);
        return Math.Clamp(pct, 0, 100);
    }
}
